flow/layers/physical_shape_layer.cc

```diff
@@ -127,47 +135,16 @@ void PhysicalShapeLayer::Paint(PaintContext& context) const {
   }
 }
 
-SkRect PhysicalShapeLayer::ComputeShadowBounds(const SkRect& bounds,
+SkRect PhysicalShapeLayer::ComputeShadowBounds(const SkPath& path,
                                                float elevation,
-                                               float pixel_ratio) {
-  // The shadow offset is calculated as follows:
-  //                   .---                           (kLightRadius)
-  //                -------/                          (light)
-  //                   |  /
-  //                   | /
-  //                   |/
-  //                   |O
-  //                  /|                              (kLightHeight)
-  //                 / |
-  //                /  |
-  //               /   |
-  //              /    |
-  //             -------------                        (layer)
-  //            /|     |
-  //           / |     |                              (elevation)
-  //        A /  |     |B
-  // ------------------------------------------------ (canvas)
-  //          ---                                     (extent of shadow)
-  //
-  // E = lt        }           t = (r + w/2)/h
-  //                } =>
-  // r + w/2 = ht  }           E = (l/h)(r + w/2)
-  //
-  // Where: E = extent of shadow
-  //        l = elevation of layer
-  //        r = radius of the light source
-  //        w = width of the layer
-  //        h = light height
-  //        t = tangent of AOB, i.e., multiplier for elevation to extent
-  // tangent for x
-  double tx =
-      (kLightRadius * pixel_ratio + bounds.width() * 0.5) / kLightHeight;
-  // tangent for y
-  double ty =
-      (kLightRadius * pixel_ratio + bounds.height() * 0.5) / kLightHeight;
-  SkRect shadow_bounds(bounds);
-  shadow_bounds.outset(elevation * tx, elevation * ty);
-
+                                               SkScalar dpr,
+                                               const SkMatrix& ctm) {
+  SkRect shadow_bounds(path.getBounds());
+  SkShadowUtils::GetLocalBounds(
+      ctm, path, SkPoint3::Make(0, 0, dpr * elevation),
+      SkPoint3::Make(kLightXOffset, kLightYOffset, dpr * kLightHeight),
+      dpr * kLightRadius, SkShadowFlags::kDirectionalLight_ShadowFlag,
+      &shadow_bounds);
```

Comment:
I'm glad to see this switch to using the Shadow utils. I discovered that the previous implementation of PSL:CSB was not producing the right values when I was doing bounds testing for the new DisplayList code, but hadn't had a chance to come back and see what the problem was.

A descriptive comment here on why and where the DPR is being factored in would be nice. In my brief investigation I could see that the matrix had DPR built in for X and Y, but there was no scale for Z so the code here seems to be manually scaling the Z components of the lighting equations by that factor. One question is that the LightOffset parameters are not being scaled by anything, but perhaps the ShadowUtils will interpret them relative to the CTM being supplied. Is that then consistent with modifying kLightHeight by DPR, but not the other values? Also, the light radius is in the X & Y directions, so does it produce the right behavior when multiplied by DPR?

Have you run tests that show that these equations produce the right values as the widget is scaled? (in other words, the shadow grows or does not grow as expected?)

Reply:
Also, if you are rebased to a recent ToT then you should see `flow/display_list_utils.cc` with calculations to compute the bounds of a shadow. It would be nice to double check it and make it consistent with the work being done here.

Reply:
I didn't thought this. I just copied the old implementation. Now I'm going to test that `ComputeShadowBounds` calculates the right values between different `dpr` values, and add that test to unit tests.

Also I will update `display_list_utils.cc`.